File: erpnext/manufacturing/report/work_order_report/work_order_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

##########Created by Cheten Tshering on 14/09/2020 ###########################
from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters):
	if filters.get("report_type") == "In_Progress":
		columns = get_progress_columns(filters)
		data = get_progress_data(filters)
	else:
		columns = get_completed_columns(filters)
		data = get_completed_data(filters)
	return columns, data

def get_progress_columns(filters=None):
	columns = [
		 	_("Date ") + ":Date:120",
            _("Work Order.") + ":Link/Work Order:120",
            _("Item Code") + ":Link/Item:100",
            _("Item Name") + ":Data:170",
			_("Item Sub Group type") + ":Link/Item Sub Group Type:150",
            _("Total") + ":Data:120",
            _("In Progress") + ":Data:120",
            _("Completed") + ":Data:90"
	]
	return columns

def get_progress_data(filters):
	query = """select 
					wo.creation
					, wo.name
					, wo.production_item
					, wo.item_name
					,i.item_sub_group_type
					, wo.qty
					, (wo.qty-wo.produced_qty) as progress
					, wo.produced_qty
         from `tabWork Order` wo, `tabItem` i
		 where wo.docstatus =1 and wo.status ='In Process' and wo.qty != wo.produced_qty and i.name=wo.production_item
		  """
		#wo.status != 'Stopped'
	if filters.get("from_date") and filters.get("to_date"):
		query += " and wo.creation between \'" + str(filters.from_date) + "\' and \'"+ str(filters.to_date) + "\'"
	return frappe.db.sql(query)

def get_completed_columns(filters=None):
	columns=[
		_("Work Order") + ":Link/Work Order:200",
		_("Date") + ":Date:120",
		_("Item") + ":Link/Item:150",
		_("Item Name") + ":Link/Item:150",
		_("Item Sub Group Type") + ":Link/Item Sub Group Type:150",
		_("To Produce") + ":Int:100",
		_("Produced") + ":Int:100",
		_("COP") + ":Float:140",
		_("Job_No") + "::80"
	]
	return columns

def get_completed_data(filters=None):
	data = frappe.db.sql("""
		select 
			wo.name
			, wo.creation
			, wo.production_item
			, wo.item_name
			, i.item_sub_group_type
			, wo.qty
			, wo.produced_qty
			, bom.total_cost
			, wo.job_no
		from `tabWork Order` wo
			left join `tabBOM` as bom on bom.name = wo.bom_no
			join `tabItem` i on i.name=wo.production_item
		where wo.docstatus = 1 and wo.status='Completed' and wo.produced_qty=wo.qty 	
		
	""")
	return data
# join `tabItem` i on i.name=wo.production_item
# 	SELECT
#   `tabWork Order`.name as "Work Order:Link/Work Order:200",
#   `tabWork Order`.creation as "Date:Date:120",
#   `tabWork Order`.production_item as "Item:Link/Item:150",
#   `tabWork Order`.item_name as "Item Name:Link/Item:150",
#   `tabWork Order`.qty as "To Produce:Int:100",
#   `tabWork Order`.produced_qty as "Produced:Int:100",
#    (select total_cost from `tabBOM` where name = `tabWork Order`.bom_no) as "COP:Float:140"
# FROM
#   `tabWork Order`
# WHERE
#   `tabWork Order`.docstatus=1
#   AND ifnull(`tabWork Order`.produced_qty,0) = `tabWork Order`.qty

File: erpnext/manufacturing/report/work_order_report/work_order_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Work Order Report"] = {
	"filters": [
		{
			"fieldname": "report_type",
			"label": "Report Type",
			"fieldtype": "Select",
			"options": ["In_Progress", "Completed"],
			"reqd": 1,
			"default": "In_Progress",
			"on_change": function (query_report) {
				var report_type = query_report.get_values().report_type;
				var from_date_filter = query_report.filters_by_name["from_date"];
				var to_date_filter = query_report.filters_by_name["to_date"];

				if (report_type == 'Completed') {
					from_date_filter.toggle(false);
					to_date_filter.toggle(false);
					from_date_filter.df.reqd = 0;
					to_date_filter.df.reqd = 0;
				}
				else {
					from_date_filter.toggle(true);
					to_date_filter.toggle(true);
					from_date_filter.df.reqd = 1;
					to_date_filter.df.reqd = 1;
				}
				query_report.refresh();
				from_date_filter.refresh();
				to_date_filter.refresh();

			}
		},
		{
			"fieldname":"from_date",
			"label": ("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1
		},
		{
			"fieldname":"to_date",
			"label": ("To Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1
		},
		{
			"fieldname": "cost_center",
			"label": ("Parent Branch"),
			"fieldtype": "Link",
			"options": "Cost Center",
			"get_query": function() {
				return {
					'doctype': "Cost Center",
					'filters': [
						['is_disabled', '!=', '1'],
						['is_group', '=', '1']
					]
				}
			},
			// "on_change": function(query_report) {
			//         var cost_center = query_report.get_values().cost_center;
			//         query_report.filters_by_name.branch.set_input(null);
			//         query_report.filters_by_name.location.set_input(null);
			//         query_report.trigger_refresh();
			//         if (!cost_center) {
			//                 return;
			//         }
			// frappe.call({
			//                 method: "erpnext.custom_utils.get_branch_from_cost_center",
			//                 args: {
			//                         "cost_center": cost_center,
			//                 },
			//                 callback: function(r) {
			//                         query_report.filters_by_name.branch.set_input(r.message)
			//                         query_report.trigger_refresh();
			//                 }
			//         })
			// },
		},
		{
		"fieldname": "branch",
		"label": ("Branch"),
		"fieldtype": "Link",
		"options": "Cost Center",
		"get_query": function() {
			var cost_center = frappe.query_report.filters_by_name.cost_center.get_value();
			// var company = frappe.query_report.filters_by_name.company.get_value();
			if(cost_center!= 'Natural Resource Development Corporation Ltd - NRDCL')
			{
				return {"doctype": "Cost Center", "filters": {"is_disabled": 0, "parent_cost_center": cost_center}}
			}
			else
			{
				return {"doctype": "Cost Center", "filters": {"is_disabled": 0, "is_group": 0}}
			}
		}
	},
	{
		"fieldname": "item",
		"label": ("Item"),
		"fieldtype": "Link",
		"options": "Item",
		"get_query": function() {
			return {
				'doctype': "Item",
				'filters': [
					['disabled', '!=', '1'],
				]
			}
		},
	}
	]
}
